[INTEGRATION][AIRFLOW] Introduce a new extractor for SFTPOperator

sekikn opened the pull request:
Signed-off-by: Kengo Seki <sekikn@apache.org>

<!-- SPDX-License-Identifier: Apache-2.0 -->

### Problem

Currently, OL doesn't have an extractor for tracing file transfer between local file systems.

### Solution

This PR introduces a new extractor for SFTPOperator for the purpose described above.

- [ ] Your change modifies the [core](https://github.com/OpenLineage/OpenLineage/blob/main/spec/OpenLineage.json) OpenLineage model
- [ ] Your change modifies one or more OpenLineage [facets](https://github.com/OpenLineage/OpenLineage/tree/main/spec/facets)

If you're contributing a new integration, please specify the scope of the integration and how/where it has been tested (e.g., Apache Spark integration supports `S3` and `GCS` filesystem operations, tested with AWS EMR).

### Checklist

- [x] You've [signed-off](https://github.com/OpenLineage/OpenLineage/blob/main/why-the-dco.md) your work
- [x] Your pull request title follows our [guidelines](https://github.com/OpenLineage/OpenLineage/blob/main/CONTRIBUTING.md#creating-pull-requests)
- [x] Your changes are accompanied by tests (_if relevant_)
- [x] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [x] You've updated any relevant documentation (_if relevant_)
- [ ] You've updated the [`CHANGELOG.md`](https://github.com/OpenLineage/OpenLineage/blob/main/CHANGELOG.md) with details about your change under the "Unreleased" section (_if relevant, depending on the change, this may not be necessary_)
- [ ] You've versioned the core OpenLineage model or facets according to [SchemaVer](https://docs.snowplowanalytics.com/docs/pipeline-components-and-applications/iglu/common-architecture/schemaver) (_if relevant_)
- [ ] You've added a [header](https://github.com/OpenLineage/OpenLineage/tree/main/.github/header_templates.md) to source files (_if relevant_)

JDarDagran reviewed (commented):
@sekikn thanks for your PR! Tests are breaking but those got fixed. Please rebase in order to fix them in your branch.

sekikn commented on integration/airflow/openlineage/airflow/extractors/sftp_extractor.py:
```diff
@@ -0,0 +1,73 @@
+            remote_host = self.operator.remote_host
+        elif hasattr(self.operator, "sftp_hook") and self.operator.sftp_hook is not None:
+            remote_host = self.operator.sftp_hook.get_connection().host
+        elif hasattr(self.operator, "ssh_hook") and self.operator.ssh_hook is not None:
```
> SFTPOperator always has remote_host, sftp_hook, ssh_hook and I don't think this is needed to use hasattr in each case.

I don't think that it's always true, for example in case that only `ssh_conn_id` is provided and neither `sftp_hook` nor `ssh_hook` is specified by user. In such case, [both hook variables are `None` within `SFTPOperator`'s constructor](https://github.com/apache/airflow/blob/providers-sftp/4.1.0/airflow/providers/sftp/operators/sftp.py#L129-L146) and then [`sftp_hook` is set in its `execute` method](https://github.com/apache/airflow/blob/providers-sftp/4.1.0/airflow/providers/sftp/operators/sftp.py#L151-L158). So if we try to get `remote_host` from `sftp_hook` or `ssh_hook` in `SFTPExtractor.extract()`, those attributes are `None` and `AttributeError` is raised.

I also tried to use `BaseExtractor`'s `extract_on_complete` method instead of `extract` so as to access `sftp_hook` set during `SFTPOperator.execute()` via the `task_instance` parameter, but its `task` attribute doesn't seem to be set always, as commented [here](https://github.com/apache/airflow/blob/providers-sftp/4.1.0/airflow/models/taskinstance.py#L517). `self.operator` also seems to keep the default state even in `extract_on_complete`, so I chose the current implementation.

> Also there's a logic in operator that sets sftp_hook with SFTPHook(ssh_hook=self.ssh_hook).

That logic is executed only when `ssh_hook` is explicitly provided by user.
https://github.com/apache/airflow/blob/providers-sftp/4.1.0/airflow/providers/sftp/operators/sftp.py#L134-L146

JDarDagran replied:
> I don't think that it's always true, for example in case that only ssh_conn_id is provided and neither sftp_hook nor ssh_hook is specified by user. In such case, [both hook variables are None within SFTPOperator's constructor](https://github.com/apache/airflow/blob/providers-sftp/4.1.0/airflow/providers/sftp/operators/sftp.py#L129-L146) and then [sftp_hook is set in its execute method](https://github.com/apache/airflow/blob/providers-sftp/4.1.0/airflow/providers/sftp/operators/sftp.py#L151-L158). So if we try to get remote_host from sftp_hook or ssh_hook in SFTPExtractor.extract(), those attributes are None and AttributeError is raised.

You're right. I'm sorry for pointing it out incorrectly. That's still quite odd that `sftp_hook` might be set in two different places.

> I also tried to use BaseExtractor's extract_on_complete method instead of extract so as to access sftp_hook set during SFTPOperator.execute() via the task_instance parameter, but its task attribute doesn't seem to be set always, as commented [here](https://github.com/apache/airflow/blob/providers-sftp/4.1.0/airflow/models/taskinstance.py#L517). self.operator also seems to keep the default state even in extract_on_complete, so I chose the current implementation.

+1

>>Also there's a logic in operator that sets sftp_hook with SFTPHook(ssh_hook=self.ssh_hook).
>
> That logic is executed only when ssh_hook is explicitly provided by user.
> https://github.com/apache/airflow/blob/providers-sftp/4.1.0/airflow/providers/sftp/operators/sftp.py#L134-L146

I meant here to skip checking if `ssh_hook` is set as `sftp_hook` overtakes it (`SFTPHook` accepts `ssh_hook` argument - it's marked deprecated anyway). However, we could assume we don't know how operator's code will change. Let's keep it as it is now just to be safe.

sekikn commented on integration/airflow/openlineage/airflow/extractors/sftp_extractor.py:
```diff
@@ -0,0 +1,73 @@
+        if isinstance(self.operator.remote_filepath, str):
+            remote_filepath = [self.operator.remote_filepath]
+        else:
+            remote_filepath = self.operator.remote_filepath
```
That's true, thank you for pointing it out! I'll remove these unnecessary if-statements.

sekikn commented on spec/Naming.md:
```diff
@@ -239,6 +239,18 @@ Identifier :
+   * Authority = {host}
+ * Unique name: {path}
+   * URI = file://{host}{path}
+
```
> I wonder if adding port wouldn't be suitable in cases like load balancer or NAT forwarder.

Makes sense for the remote port, but I think the local port is not determined. Is it OK to adopt only `{host}` for the local file and `{host}:{port}` for the remote files as authority?

> Also this is under discussion what exact convention should be established for local files. Please keep in mind that above might slightly change in near future.

That's OK to me, thanks for notifying!

JDarDagran replied:
Should be with port in all 3 llines for namespace, authority and URI.

sekikn commented:
Thank you for reviewing and addressing the mypy issue, @JDarDagran! I left some replies.

The latest CI failures are caused by timeout and the Spark IT, so it seems to be irrelevant to this PR.

sekikn commented:
Thank you for the response @JDarDagran, just updated the PR. Let's see how CI goes.

> On secound thought I think we should change convention to use IP instead of host name. Generally speaking we should aim into making dataset namespace + name unique. Use of IP makes it possible in most cases.

Regarding this, I fixed the latest PR as follows, considering a case using dummy hostname (e.g., unit test):

* If hostname can be resolved to IP address, use it
* If not, still use hostname but leave a warning in log

sekikn commented:
Oh, I've just noticed I forgot to fix the integration test. It will fail since I didn't change requests/sftp.json at all.
I've also realized that it may be difficult to fix IP addresses for integration test in the CI environment which is built with docker-compose. Anyway I'll try it.

sekikn commented:
I fixed the IP issue in the latest PR. The following document helped me.
https://docs.docker.com/compose/compose-file/compose-file-v2/#ipv4_address-ipv6_address

The most recent CI run failed due to timeout in the previous step in question (it seems to frequently happen), but I ensured the latest PR passed the integration test locally.

```
$ git show --shortstat 
commit 9543ad723a1f644fcc3d7e9aafd27cfd691caa5f (HEAD -> feature/sftp-extractor, origin/feature/sftp-extractor)
Author: Kengo Seki <sekikn@apache.org>
Date:   Mon Nov 14 08:51:09 2022 +0900

    Fix IP address of worker and SSH server for integration test
    
    Signed-off-by: Kengo Seki <sekikn@apache.org>

 2 files changed, 37 insertions(+), 6 deletions(-)
$ AIRFLOW_IMAGE=apache/airflow:2.4.2-python3.7 ./tests/integration/docker/up.sh

...

test_integration.py::test_integration[sftp_dag-requests/sftp.json]                                                                                                           
------------------------------------------------------------------------------- live log call -------------------------------------------------------------------------------
2022-11-14 01:19:32 INFO Checking dag sftp_dag                                                                                                                               
2022-11-14 01:19:32 INFO Waiting for DAG 'sftp_dag'...                                                                                                                       
2022-11-14 01:19:32 INFO DAG 'sftp_dag' state set to 'success'.                                                                                                              
2022-11-14 01:19:37 DEBUG Starting new HTTP connection (1): backend:5000                                                                                                     
2022-11-14 01:19:37 DEBUG http://backend:5000 "GET /api/v1/lineage HTTP/1.1" 200 208350                                                                                      
2022-11-14 01:19:37 INFO failed to compare events for dag sftp_dag!                                                                                                          
2022-11-14 01:19:37 INFO Events for dag sftp_dag verified!                                                                                                                   
PASSED                                                                                                                                                                [ 82%] 

...

================================================================= 13 passed, 4 skipped in 89.63s (0:01:29) ==================================================================
```

@JDarDagran Would you take a look?

JDarDagran commented:
@sekikn I've made some improvements in the CI process. Please rebase once again so that tests should run successfully.

The one thing that I made notice previously and it actually just happens. As a result of a broader discussion - `{host}:{port}` should be the only choice without having port optional. In case of local files we do not know remote port let's default to `paramiko.config.SSH_PORT` value. Besides it looks good! Thank you!

sekikn commented:
Thanks a lot for addressing the CI issue, @JDarDagran! I'll update the PR in accordance with your comment.

There's one more thing I've just noticed, the current implementation doesn't pass the integration test on CI because of the dependency version. [Airflow's dependencies are installed using the constraint file](https://github.com/OpenLineage/OpenLineage/blob/main/integration/airflow/setup.cfg#L38), and it installs apache-airflow-providers-sftp 2.1.1 for Airflow 2.1.4, sftp 2.4.1 for Airflow 2.2.4, and sftp 4.0.0 for Airflow 2.3.4 respectively, while this extractor requires sftp 4.1.0+. We can add a new test condition for addressing this, but I'd rather loosen the version requirement by accepting both str and list as `local_filepath` and `remote_filepath` so that wider range of users can leverage this extractor.

JDarDagran commented:
Let's loosen the requierements.

JDarDagran commented on integration/airflow/openlineage/airflow/extractors/sftp_extractor.py:
```diff
@@ -0,0 +1,109 @@
+        )
+
+    def _get_source(self, scheme, host, port, path) -> Source:
+        authority = host if port is None else f"{host}:{port}"
```
`port` should not be optional in any case, even for local files. It should be:
```
port = port or paramiko.config.SSH_PORT
authority = f"{host}:{port}"
```

JDarDagran commented on integration/airflow/openlineage/airflow/extractors/sftp_extractor.py:
```diff
@@ -0,0 +1,109 @@
+        elif hasattr(hook, "ssh_hook"):
+            remote_port = hook.ssh_hook.port
+        if remote_port is None:
+            remote_port = paramiko.config.SSH_PORT
```
If we change `_get_source` this should be removed here.

JDarDagran commented on integration/airflow/tests/integration/requests/sftp.json:
```diff
@@ -0,0 +1,42 @@
+                }
+            },
+            "name": "/opt/airflow/airflow.cfg",
+            "namespace": "file://172.16.238.100"
```
This should be changed to `172.16.238.100:22`

JDarDagran reviewed (commented):
@sekikn thank you again for your effort! I might have been misunderstood. Having a mandatory port, we should also set a default port for local files.

sekikn commented:
Oh, sorry for my misunderstanding. Now I got what you meant, thank you for the elaboration!

JDarDagran reviewed (approved):
@sekikn LGTM! Thank you a lot!
